// docpad-plugin-coffeescript/out/coffeescript.plugin.js
// Generated by CoffeeScript 1.9.1
(function() {
  var extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
    hasProp = {}.hasOwnProperty;

  module.exports = function(BasePlugin) {
    var CoffeescriptPlugin;
    return CoffeescriptPlugin = (function(superClass) {
      extend(CoffeescriptPlugin, superClass);

      function CoffeescriptPlugin() {
        return CoffeescriptPlugin.__super__.constructor.apply(this, arguments);
      }

      CoffeescriptPlugin.prototype.name = 'coffeescript';

      CoffeescriptPlugin.prototype.config = {
        compileOptions: {}
      };

      CoffeescriptPlugin.prototype.render = function(opts) {
        var coffee, compileOptions, compiled, err, file, fileFullPath, finish, inExtension, js, key, literate, outExtension, ref, start, v3SourceMap, value;
        inExtension = opts.inExtension, outExtension = opts.outExtension, file = opts.file;
        literate = false;
        if (((inExtension === 'coffee' || inExtension === 'litcoffee') && (outExtension === 'js' || outExtension === null)) || ((inExtension === 'md' || inExtension === 'markdown') && outExtension === 'js' && (literate = true))) {
          coffee = require('coffee-script');
          fileFullPath = file.get('fullPath');
          compileOptions = {
            filename: fileFullPath,
            sourceFiles: [file.get('url') + '.coffee'],
            generatedFile: file.get('url'),
            literate: literate || coffee.helpers.isLiterate(fileFullPath)
          };
          ref = this.getConfig().compileOptions;
          for (key in ref) {
            if (!hasProp.call(ref, key)) continue;
            value = ref[key];
            if (compileOptions[key] == null) {
              compileOptions[key] = value;
            }
          }
          try {
            compiled = coffee.compile(opts.content, compileOptions);
            if (typeof compiled === 'object') {
              js = compiled.js, v3SourceMap = compiled.v3SourceMap;
            } else {
              js = compiled;
            }
            if (v3SourceMap) {
              v3SourceMap = JSON.parse(v3SourceMap);
              v3SourceMap.sourcesContent = [opts.content];
              v3SourceMap = JSON.stringify(v3SourceMap, null, 2);
              js = "" + js + "\n//# sourceMappingURL=data:application/json;base64," + new Buffer(unescape(encodeURIComponent(v3SourceMap)), 'binary').toString('base64');
            }
            opts.content = js;
          } catch (_error) {
            err = _error;
            if (err.location) {
              start = err.location.first_line + ":" + err.location.first_column;
              finish = err.location.last_line + ":" + err.location.last_column;
              err.message += " at " + fileFullPath + ":" + start;
              if (finish !== start) {
                err.message += " to " + finish;
              }
            }
            return err;
          }
        }
      };

      return CoffeescriptPlugin;

    })(BasePlugin);
  };

}).call(this);
